fc_util/src/arg_parser.rs

```diff
@@ -0,0 +1,581 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Borrow;
+use std::collections::HashSet;
+use std::env;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::process;
+use std::result;
+
+pub type Result<T> = result::Result<T, Error>;
+
+/// Errors associated with parsing and validating arguments.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// The required argument was not provided.
+    MissingArgument(String),
+    /// A value for the argument was not provided.
+    MissingValue(String),
+    /// The provided argument was not expected.
+    UnexpectedArgument(String),
+    /// The argument was provided more than once.
+    DuplicatedArgument(String),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            MissingArgument(ref arg) => write!(f, "Argument '{}' required, but not found.", arg),
+            MissingValue(ref arg) => write!(
+                f,
+                "The argument '{}' requires a value, but none was supplied.",
+                arg
+            ),
+            UnexpectedArgument(ref arg) => write!(
+                f,
+                "Found argument '{}' which wasn't expected, or isn't valid in this context.",
+                arg
+            ),
+            DuplicatedArgument(ref arg) => {
+                write!(f, "The argument '{}' was provided more than once.", arg)
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct ArgInfo<'a> {
+    name: &'a str,
+    required: bool,
+    requires: Option<&'a str>,
+    takes_value: bool,
+    default_value: Option<&'a str>,
+    optional: bool,
+    help: Option<&'a str>,
+    user_value: Option<String>,
+}
+
+impl<'a> ArgInfo<'a> {
+    pub fn new(name: &'a str) -> ArgInfo<'a> {
+        ArgInfo {
+            name,
+            required: false,
+            requires: None,
+            takes_value: false,
+            default_value: None,
+            optional: false,
+            help: None,
+            user_value: None,
+        }
+    }
+
+    pub fn required(mut self, required: bool) -> ArgInfo<'a> {
+        self.required = required;
+        self
+    }
+
+    pub fn requires(mut self, requires: &'a str) -> ArgInfo<'a> {
+        self.requires = Some(requires);
+        self
+    }
+
+    pub fn takes_value(mut self, takes_value: bool) -> ArgInfo<'a> {
+        self.takes_value = takes_value;
+        self
+    }
+
+    pub fn default_value(mut self, default_value: &'a str) -> ArgInfo<'a> {
+        self.default_value = Some(default_value);
+        self
+    }
+
+    pub fn optional(mut self, optional: bool) -> ArgInfo<'a> {
+        self.optional = optional;
+        self
+    }
+
+    pub fn help(mut self, help: &'a str) -> ArgInfo<'a> {
+        self.help = Some(help);
+        self
+    }
+
+    pub fn update_value(mut self, value: String) -> ArgInfo<'a> {
```

Comment:
I would rename this to "set_user_value". Also is it possible to receive a `&'a str` here instead of a `String` ?

Reply:
I meant holding this struct member (`user_value`) as a &'a str . Just like all the other. I don't know if it makes sense.

Reply:
As per our chat discussion, we've decided to keep `String` here because of the lifetimes conflicts that would appear.